### PR TITLE
Little web worker framework

### PR DESCRIPTION
**Pull request by rpiaggio**
This PR introduces a layer to abstract as much as possible the use of web workers.

All communication between UI and workers is done via a low level protocol that keeps track of the process that initiated requests and routes responses accordingly, muxing/demuxing messages through the worker's channel.

Basically, a client makes a request and receives a `Resource[Stream]` of responses. The client can cancel the request (by releasing the resource, and this will cancel the running fiber in the server. Server errors are propagated to the client's stream.

Convenience methods are provided to request only one element, in which case a `F[A]` is returned; or to post a message to a server without awaiting response, in which case a `F[Unit]` is returned.

---

This PR also allows defining multiple workers without each one being bundled with the whole module it depends on. This is achieved by using dynamic imports from a thin `js` file for the worker.

The previous worker was refactored into 3 separate workers, each in its own thread; namely:
- ITC
- Guide Star Catalogue
- AGS

---

Finally, utils were rearranged a bit.

### CODE REVIEW

File: explore/src/main/webapp/agsworker.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import { AgsWorker } from '@workers/exploreworkers1.js';
```
Comment (cquiroz) [inline explore/src/main/webapp/agsworker.js:1]: Why `exploreworkers1`?

Comment (rpiaggio) [reply explore/src/main/webapp/agsworker.js:1]: Just testing, all workers should be in the same module now.

File: vite.config.js [hunk 2]
```diff
@@ -178,10 +178,11 @@ module.exports = ({ command, mode }) => {
       },
       outDir: path.resolve(__dirname, 'heroku/static'),
     },
+    worker: {
+      format: 'es'
```
Comment (cquiroz) [inline vite.config.js:182]: I don't recall this before, it is because the dynamic import?

File: vite.config.js [hunk 3]
```diff
@@ -178,10 +178,11 @@ module.exports = ({ command, mode }) => {
...
+    },
     plugins: [
-      isProduction
```
Comment (cquiroz) [inline vite.config.js:182]: I think we may want to restore this

File: workers/src/main/scala/workers/WorkerServer.scala [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  /**
+   * Handle server-specific messages.
+   */
+  protected def handler: Logger[F] ?=> F[Invocation => F[Unit]]
+
+  private val F = summon[Sync[F]]
+
+  protected def setupLogger: F[Logger[F]] = Sync[F].delay {
```
Comment (cquiroz) [inline workers/src/main/scala/workers/WorkerServer.scala:55]: I guess this is going to be a global setting. do we need to initialize it for each worker?

File: workers/src/main/scala/workers/WorkerServer.scala [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                              Invocation(
+                                data,
+                                pickled =>
+                                  postAsTransferable[F, FromServer](self, FromServer.Data(id, pickled))
+                              )
+                            )
+                    _    <- postAsTransferable[F, FromServer](self, FromServer.Complete(id))
+                  } yield ()
+                ).handleErrorWith(t =>
+                  postAsTransferable[F, FromServer](
```
Comment (cquiroz) [inline workers/src/main/scala/workers/WorkerServer.scala:85]: I guess this means we send errors back to the client, right?

**Review by cquiroz — APPROVED** [review]
This is looking great. Looking forward to test it

File: common/src/main/scala/explore/model/WorkerClients.scala [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+object WorkerClients {
+
+  /**
+   * This deserves an explanation:
+   *
+   * To make the webworker act correctly in both dev and production we shoud import it as a module
+   * rather than just doing a direct consructor call.
+   *
+   * Doing the import with the "worker" param gives a constructor for the worker which we can wrap
+   * inline lets us save some space keeping a single chunk More info see:
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply common/src/main/scala/explore/model/WorkerClients.scala:33]: :+1:

File: common/src/main/scala/workers/WorkerClient.scala [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+class WorkerClient[F[_]: Concurrent: UUIDGen, R: Pickler](worker: WebWorkerF[F]) {
+
+  /**
+   * Make a request to the underlying worker and receive responses as a `Stream`.
+   */
+  def request[T <: R & WorkerRequest](requestMessage: T)(using
+    Pickler[requestMessage.ResponseType]
+  ): Resource[F, fs2.Stream[F, requestMessage.ResponseType]] =
+    for {
+      id     <- Resource.eval(UUIDGen.randomUUID).map(WorkerProcessId.apply)
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply common/src/main/scala/workers/WorkerClient.scala:30]: Yes, all messages are demuxed with a UUID, which is generated on each request.

File: common/src/main/scala/workers/WorkerClient.scala [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Implements the client side of a simple client/server protocol that provides a somewhat more
+ * functional/effecful way of communicating with workers.
+ */
+class WorkerClient[F[_]: Concurrent: UUIDGen, R: Pickler](worker: WebWorkerF[F]) {
+
+  /**
+   * Make a request to the underlying worker and receive responses as a `Stream`.
+   */
+  def request[T <: R & WorkerRequest](requestMessage: T)(using
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply common/src/main/scala/workers/WorkerClient.scala:26]: The `Pickler[R]` will be used for `requestMessage`. This has to be the same type defined in the `WorkerServer`. Pickling requires we pickle and unpickle with the same supertype, even when using subtypes (ie: if `T <: R`,  a `Pickler[R]` won't unpickle something pickler with `Pickler[T]`, even if `Pickler[R]` is a composite pickler built with the same `Pickler[T]`).

File: common/src/main/scala/workers/WorkerClient.scala [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+      .collect {
+        case FromServer.Data(mid, pickled) if mid === id =>
+          fromBytes[requestMessage.ResponseType](pickled.value).some
+        case FromServer.Complete(mid) if mid === id      =>
+          none
+        case FromServer.Error(mid, error) if mid === id  =>
+          throw error
+      }
+      .unNoneTerminate
+      .rethrow
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply common/src/main/scala/workers/WorkerClient.scala:49]: Yes, that's the idea

File: common/src/main/scala/workers/WorkerClient.scala [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * Make a request to the underlying worker and receive a single response (if any) as the effect
+     * result.
+     */
+  def requestSingle[T <: R & WorkerRequest](requestMessage: T)(using
+    Pickler[requestMessage.ResponseType]
+  ): F[Option[requestMessage.ResponseType]] = // TODO Should we implement a timeout here? Retry?
+    request(requestMessage).use(_.head.compile.last)
+
+  given Pickler[Nothing] = ??? // Intentionally left unimplemented, should never be called
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply common/src/main/scala/workers/WorkerClient.scala:60]: This pickler will never be called, but it's needed to compile `invoke` (which sends a message to the server but doesn't wait for a return message).

I'll change to Pickler that throws an exception if invoked. I'll also rename `invoke` to `requestAndFoget`.

File: explore/src/main/scala/explore/config/ConfigurationPanel.scala [hunk 11]
```diff
@@ -131,32 +132,18 @@ object ConfigurationPanel {
...
+          .requestSingle(
+            ItcMessage.SpectroscopyMatrixRequest(uri"/instrument_spectroscopy_matrix.csv")
           )
-          .whenA(mx === PotOption.ReadyNone)
       }
       .render { (props, editState, matrix) =>
         implicit val ctx: AppContextIO = props.ctx
 
-        implicit val client = ctx.clients.odb // This shouldn't be necessary, but it seems to be
+        // implicit val client = ctx.clients.odb // This shouldn't be necessary, but it seems to be
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply explore/src/main/scala/explore/config/ConfigurationPanel.scala:146]: Lol, yes! It's not necessary anymore.

File: model-tests/shared/src/test/scala/explore/model/boopickle/BoopickleSuite.scala [hunk 12]
```diff
@@ -59,7 +57,4 @@ class BoopickleSuite
   checkAll("Pickler[Wavelengtth]", PicklerTests[Wavelength].pickler)
 
   checkAll("Pickler[UnnormalizedSED]", PicklerTests[UnnormalizedSED].pickler)
-
-  // Events
-  checkAll("Pickler[CatalogResults]", PicklerTests[CatalogResults].pickler)
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply model-tests/shared/src/test/scala/explore/model/boopickle/BoopickleSuite.scala:64]: There's no more `CatalogResults`. Now we return directly the type it used to wrap `List[GuideStarCandidate]`: https://github.com/gemini-hlsw/explore/blob/workers/model/shared/src/main/scala/explore/events/CatalogMessage.scala#L23

File: workers/src/main/scala/workers/WorkerServer.scala (hunk 5 above)
Comment (rpiaggio) [reply workers/src/main/scala/workers/WorkerServer.scala:85]: Yes. In order to be able to pickle exceptions we transform them into `WorkerException`.

File: workers/src/main/scala/workers/WorkerServer.scala (hunk 4 above)
Comment (rpiaggio) [reply workers/src/main/scala/workers/WorkerServer.scala:55]: Web workers cannot share things AFAIK, they each run in its own isolated thread. So they need to create a new logger each. A plus is that logging can be tweaked independently for each worker by overriding this.

File: vite.config.js (hunk 3 above)
Comment (rpiaggio) [reply vite.config.js:182]: The "production" setting here only seems to apply when running within vite in production mode, that is with "npm run serve".  This is convenient when we want to locally test the production build.

It doesn't affect our deployment at all, which is done with static files.

File: vite.config.js (hunk 2 above)
Comment (rpiaggio) [reply vite.config.js:182]: Yes, I'll add a comment

File: explore/src/main/webapp/project/build.properties [hunk 13]
```diff
@@ -0,0 +1 @@
+sbt.version=1.6.2
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply explore/src/main/webapp/project/build.properties:1]: I'm very curious as to what generated this. I'll try removing.

File: explore/src/main/webapp/agsworker.js [hunk 14]
```diff
@@ -0,0 +1,4 @@
+(async () => {
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply explore/src/main/webapp/agsworker.js:1]: See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import#import_a_module_for_its_side_effects_only

I'll add a comment anyway

File: explore/src/main/scala/explore/itc/ItcGraphPanel.scala [hunk 15]
```diff
@@ -98,49 +98,31 @@ object ItcGraphPanel {
...
-                  .postWorkerMessage(ItcGraphQuery(u, w, sn, constraints, t, mode))
-              )
-              .orEmpty).getOrElse(IO.unit).flatten
+        (for
+          w           <- OptionT.fromOption[IO](wavelength)
+          sn          <- OptionT.fromOption[IO](signalToNoise)
+          constraints <- OptionT.fromOption[IO](constraints)
+          t           <- OptionT.fromOption[IO](itcTargets)
+          mode        <- OptionT.fromOption[IO](instrumentRow)
+        yield ItcClient[IO]
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply explore/src/main/scala/explore/itc/ItcGraphPanel.scala:119]: There's a single instance of each worker and a client for each of them. Both are created at the app initialization: here https://github.com/gemini-hlsw/explore/blob/workers/explore/src/main/scala/explore/Explore.scala#L214 and here https://github.com/gemini-hlsw/explore/blob/workers/common/src/main/scala/explore/model/WorkerClients.scala#L61.

The client will assign a new UUID to each request, create a new stream from the `Topic` in `WebWorkerF`, and only process responses with the same UUID. See the if guards in the `collect` here: https://github.com/gemini-hlsw/explore/blob/workers/common/src/main/scala/workers/WorkerClient.scala#L41

**Pull request merged**